text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,641 @@
+"Receive" spans.
+
+Operations covered by "Deliver" or "Receive" can forward zero messages (e. g.
+to notify the application that no message is available for processing), one
+message, or multiple messages (a batch of messages). "Deliver" and "Receive"
+spans should link to the "Create" or "Publish" span of the messages forwarded,
+thus those spans can link to zero, one, or multiple "Create" spans.
+
+For single-message scenarios, in addition to adding the "Create" span as a link
+on the "Deliver" or "Receive" span, the "Create" span may also be used as a
+parent on those operations. Keeping single-messages operations in the same
+trace can greatly improve the user experience.
+
+#### Settlement
+
+Messages can be settled in a variety of different ways:
+
+* Messages are not settled at all (fire-and-forget).
+* Settlement happens on the broker.
+* Settlement operations are triggered manually by the user.
+* In callback scenarios settlement can be automatically triggered by messaging SDKs
+  based on return values of callbacks.
+
+A "Settle" span should be created for every settlement operation, no matter
+which party triggered it.  SDKs will, in some cases, auto-settle messages in
+push-scenarios when messages are delivered via callbacks. In cases where it is
+possible, it is recommended to create the "Settle" span as a child of the
+"Deliver" span.
+
+"Settle" spans SHOULD link to "Create" or "Publish" spans of the messages that are
+settled, when possible.
+
+## Proposed changes and additions to the messaging semantic conventions
+
+### Operation name
+
+The following operations related to messages are covered by these semantic
+conventions:
+
+| Operation name | Description |
+|----------------|-------------|
+| `publish`      | One or more messages are provided for publishing to an intermediary. |
+| `create`       | A message is created. |
+| `receive`      | One or more messages are requested by a consumer. |
+| `deliver`      | One or more messages are passed to a consumer. |
+| `settle`       | One or more messages are settled. |
+
+For further details about each of those operations refer to the [section about trace structure](#trace-structure).
+
+### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER`, if no `create` spans are present. |
+| `create`       | `PRODUCER` |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+
+The kind of `settle` spans should be set according to the [generic specification about span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind),
+e. g. it should be set to `CLIENT` if the `settle` spans models a synchronous call
+to the intermediary.
```

Comment:
Good question, I know initial thoughts were that it can either be a `CLIENT` or an `INTERNAL` span.

You are right, if settlement is done as an asynchronous operations, the wording here would recommend to set use `PRODUCER ` as span kind for the settlement span.

@lmolkova @joaopgrassi, what are your thoughts about that?

Reply:
Yeah `PRODUCER` would be somewhat strange to me as well. Maybe it has to always be `CLIENT`?